client/state/comments/reducer.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * External dependencies
+ */
+import Immutable from 'immutable';
+import { combineReducers } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import {
+	COMMENTS_RECEIVE,
+	COMMENTS_REMOVE,
+	COMMENTS_ERROR,
+	COMMENTS_COUNT_RECEIVE,
+	COMMENTS_LIKE,
+	COMMENTS_LIKE_UPDATE,
+	COMMENTS_UNLIKE,
+	COMMENTS_REQUEST,
+	COMMENTS_REQUEST_SUCCESS,
+	COMMENTS_REQUEST_FAILURE,
+	DESERIALIZE,
+	SERIALIZE,
+} from '../action-types';
+import {
+	createCommentTargetId,
+	listFindAndModify
+} from './utils';
+import {
+	PLACEHOLDER_STATE
+} from './constants';
+
+/***
+ * Creates a function that updates like count
+ * @param {Boolean} like true to like, false to unlike
+ * @returns {Function} function that updates like count for comment
+ */
+function getLikeSetter( { like = true, likeCount } ) {
+	return ( comment ) => comment.set( 'i_like', like )
+								.update( 'like_count', ( ct ) => {
+									if ( typeof likeCount !== 'undefined' ) {
+										return likeCount;
+									}
+
+									return like ? ct + 1 : ct - 1;
+								} );
+}
+
+/***
+ * Updates a specific state in the general state
+ * @param {Immutable.Map} state general state for all commentTargetId
+ * @param {Object} action action object
+ * @param {Number} action.siteId site identifier
+ * @param {Number} action.postId post identifier
+ * @param {Function|Object} updaterOrValue an updater function or a value
+ * @returns {Immutable.Map} new state
+ */
+function updateSpecificState( state, action, updaterOrValue ) {
+	const id = createCommentTargetId( action.siteId, action.postId );
+
+	if ( typeof updaterOrValue === 'function' ) {
+		return state.update( id, ( value ) => updaterOrValue( value ) );
+	}
+
+	return state.set( id, updaterOrValue );
+}
+
+/***
+ * Comments items reducer, stores a comments items Immutable.List per siteId, postId
+ * @param {Immutable.Map} state redux state
+ * @param {Object} action redux action
+ * @returns {Immutable.Map} new redux state
+ */
+export function items( state = Immutable.Map(), action ) {
+	switch ( action.type ) {
+		case COMMENTS_RECEIVE:
+			// create set of ids for faster lookup for filter later
+			const newIds = Immutable.Set( action.comments.map( comment => comment.ID ) );
+
+			// we prefer freshly retrieved data, so throw away old data
+			return updateSpecificState( state, action, ( oldSpecificState = Immutable.List() ) =>
+				oldSpecificState.filter( ( comment ) => ! newIds.has( comment.get( 'ID' ) ) )
+								.concat( Immutable.fromJS( action.comments ) )
+								.sort( ( a, b ) => new Date( b.get( 'date' ) ) - new Date( a.get( 'date' ) ) )
+			);
+
+		case COMMENTS_REMOVE:
+			return updateSpecificState( state, action, ( oldSpecificState = Immutable.List() ) =>
+				oldSpecificState.filter( ( comment ) => comment.get( 'ID' ) !== action.commentId )
+			);
+		case COMMENTS_LIKE:
+			return updateSpecificState( state, action, ( oldSpecificState = Immutable.List() ) =>
+				listFindAndModify( oldSpecificState,
```

Comment:
I can't simply use a `sentIn` because I need to know the index on which to set (the correct comment on which to update the like), to know the index i need to find it, after I find the index I `setIn`.

That logic is encapsulated in `listFindAndModify`.

I might be missing something here since I do expect Immutable.js to handle such case, so maybe you can give an example on how to achieve that?

I'm not really following your filtering suggestion, a `filter()` creates a new collection, any modifications on that collection (or it's descendant) should not modify the old collection, that's the Immutable thing, so I can't see how that can work, again, I might be missing something here so lets chat about that =)